### ISSUE TITLE
systemd-nspawn fails to start systemd >=253 in QEMU-emulated container

### ISSUE THREAD

**Issue by MichaIng** [systemd/systemd#28901]
### systemd version the issue has been seen with

254.1-2

### Used distribution

Debian 13/Trixie/Sid

### Linux kernel version used

6.4.4-3

### CPU architectures issue was seen on

x86_64

### Component

systemd, systemd-nspawn

### Expected behaviour you didn't see

When booting a Debian container image with systemd >=253 (within the container) with non-native userland architecture via `systemd-nspawn` and `qemu-user-static` QEMU-emulator, invoked via `systemd-binfmt`, I expect it to boot successfully.

### Unexpected behaviour you saw

Booting the container fails at the point where `systemd` (PID 1) is tried to e started.

Referring to this Debian bug report, while I think it is not a Debian issue in particular: https://bugs.debian.org/1050045

Referring to this solved bug in v253 where the same error message was thrown: https://github.com/systemd/systemd/issues/26474
However, our issue persists with v254 and `systemd-nspawn` grants CAP_SYS_ADMIN by default, so it seems to be slightly differently located.

### Steps to reproduce the problem

Run the following on any dpkg-based Linux distribution on `x86_64` hardware, to `debootstrap` and boot a fresh foreign architecture Debian Trixie rootfs tree with systemd init system, and needed QEMU emulator:
```sh
sudo apt -y install debootstrap dbus systemd-container qemu-user-static 
binfmt-support
sudo systemctl restart systemd-binfmt
debootstrap --arch=arm64 --variant=minbase --include=systemd-sysv trixie 
./rootfs
systemd-nspawn -bD rootfs
```
The same fails with any other foreign architecture and also with `sid` (unstable) instead of `trixie` (testing). It worked when systemd was still v252 and started failing once v253 was merged into Debian repositories.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
Spawning container rootfs on /root/rootfs.
Press Ctrl-] three times within 1s to kill container.
systemd 254.1-2 running in system mode (+PAM +AUDIT +SELINUX +APPARMOR 
+IMA +SMACK +SECCOMP +GCRYPT -GNUTLS +OPENSSL +ACL +BLKID +CURL 
+ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 
-PWQUALITY +P11KIT +QRENCODE +TPM2 +BZIP2 +LZ4 +XZ +ZLIB +ZSTD 
-BPF_FRAMEWORK -XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified)
Detected virtualization systemd-nspawn.
Detected architecture arm64.

Welcome to Debian GNU/Linux trixie/sid!

Hostname set to <VM-Trixie>.
Failed to fork off sandboxing environment for executing generators: 
Invalid argument
[!!!!!!] Failed to start up manager.
Exiting PID 1...
Container rootfs failed with error code 255.
```

**Comment by poettering** [comment]
this is something for qemu to address. It#s syscall emulation is very incomplete. They document that even. They usually add what's missing though to make common workloads work. Here you really have to talk to qemu hence.

**Comment by MichaIng** [comment]
Okay thanks for the response. I'll check back with QEMU then.

**Comment by mjt00k** [comment]
commit ca6ce62d2a437432082b5c6e5d4275d56055510f «manager: execute generators in a mount namespace "sandbox"» (v253) introduced use of CLONE_NEWNS flag in clone() when spawning generators. This clone() flag is not supported by qemu-linux-user.

**Comment by mjt0k** [comment]
commit https://github.com/systemd/systemd/commit/ca6ce62d2a437432082b5c6e5d4275d56055510f «manager: execute generators in a mount namespace "sandbox"» (v253) introduced use of CLONE_NEWNS flag in clone() when spawning generators. This clone() flag is not supported by qemu linux-user.

**Comment by mbiebl** [comment]
@mjt0k In the downstream bug report, you mentioned

```
Ok.  This is in qemu linux-user/syscall.c:

#define CLONE_IGNORED_FLAGS                     \
    (CLONE_DETACHED | CLONE_IO)

/* Flags for fork which we can implement within QEMU itself */
#define CLONE_OPTIONAL_FORK_FLAGS               \
    (CLONE_SETTLS | CLONE_PARENT_SETTID | CLONE_PIDFD | \
     CLONE_CHILD_CLEARTID | CLONE_CHILD_SETTID)

#define CLONE_INVALID_FORK_FLAGS                                        \
    (~(CSIGNAL | CLONE_OPTIONAL_FORK_FLAGS | CLONE_IGNORED_FLAGS))

This is basically it: the list of clone(2) flags supported by
qemu-user is given in CLONE_OPTIONAL_FORK_FLAGS.

systemd does this (from qemu running with QEMU_STRACE=1):

 1 clone(CLONE_NEWNS|0x11,child_stack=0x0000000000000000,parent_tidptr=0x0000000000000000,tls=0xfffffffe7fffffff,child_tidptr=0x0000000000000000) = 
-1 errno=22 (Invalid argument)

In the same qemu source we read:

/* CLONE_VFORK is special cased early in do_fork(). The other flag bits
 * have almost all been allocated. We cannot support any of
 * CLONE_NEWNS, CLONE_NEWCGROUP, CLONE_NEWUTS, CLONE_NEWIPC,
 * CLONE_NEWUSER, CLONE_NEWPID, CLONE_NEWNET, CLONE_PTRACE, CLONE_UNTRACED.
 * The checks against the invalid thread masks above will catch these.
 * (The one remaining unallocated bit is 0x1000 which used to be CLONE_PID.)
 */

In particular, CLONE_NEWNS in qemu is marked as "we cannot support".
```

Are you saying, this is "unfixable" on qemu's side?

**Comment by mjt0k** [comment]
Well, nothing is truely unfixable.  But so far, none of CLONE_NEW* flags are supported, and might need significantly different approach to whole exec/clone thing.

**Comment by mrc0mmand** [comment]
> commit [ca6ce62](https://github.com/systemd/systemd/commit/ca6ce62d2a437432082b5c6e5d4275d56055510f) «manager: execute generators in a mount namespace "sandbox"» (v253) introduced use of CLONE_NEWNS flag in clone() when spawning generators. This clone() flag is not supported by qemu linux-user.

There was actually a follow-up to that commit - a2275dcb9d48925f8192f44ec371710ce812bfce - that falls back to running the generators without sandboxing if forking with a new mount namespace fails. However, this works only if we get EPERM/EACCESS, not EINVAL as is the case here.

**Comment by yuwata** [comment]
Fix is waiting in #28954.

**Comment by greyltc** [comment]
@yuwata does https://github.com/systemd/systemd/pull/28954 belong in the host's systemd or the guest's? I've put it into my host's systemd and I'm still getting `Failed to fork off sandboxing environment for executing generators: Invalid argument`

**Comment by MichaIng** [comment]
This is all only about systemd in the guest system. The version on the host is irrelevant.

**Comment by greyltc** [comment]
> Fix is waiting in #28954.

This patch solved the issue for me.

**Comment by MichaIng** [comment]
I just verified as well that it works with systemd 254.4 on Debian Sid. Many thanks!
